Refactor Private Click Measurement support for ephemeral sessions

whsieh opened the pull request:
#### e10cb3c48fdc2ce116cf0d74cca3c96e931e7e6d
<pre>
Refactor Private Click Measurement support for ephemeral sessions
<a href="https://bugs.webkit.org/show_bug.cgi?id=253600">https://bugs.webkit.org/show_bug.cgi?id=253600</a>

Reviewed by John Wilander.

Refactor some support for the ephemeral version of Private Click Measurement (i.e., the version
currently used in Safari view controller), such that it&apos;s additionally compatible with ephemeral
browsing sessions.

* LayoutTests/http/tests/privateClickMeasurement/database-disabled-in-ephemeral-session-expected.txt: Renamed from LayoutTests/http/tests/privateClickMeasurement/conversion-disabled-in-ephemeral-session-expected.txt.
* LayoutTests/http/tests/privateClickMeasurement/database-disabled-in-ephemeral-session.html: Renamed from LayoutTests/http/tests/privateClickMeasurement/conversion-disabled-in-ephemeral-session.html.
* LayoutTests/http/tests/privateClickMeasurement/store-disabled-in-ephemeral-session-expected.txt: Removed.
* LayoutTests/http/tests/privateClickMeasurement/store-in-ephemeral-session-expected.txt: Added.
* LayoutTests/http/tests/privateClickMeasurement/store-in-ephemeral-session.html: Renamed from LayoutTests/http/tests/privateClickMeasurement/store-disabled-in-ephemeral-session.html.

Rename these tests to better reflect the new behavior in ephemeral sessions, and adjust the test
output to match.

* Source/WebCore/html/HTMLAnchorElement.cpp:
(WebCore::HTMLAnchorElement::parsePrivateClickMeasurement const):

Remove the early return here, and continue on call out to the client layer with PCM attribution
measurement data.

* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::willSendRedirectedRequestInternal):

Additionally avoid going through with attribution, if the click measurement happened too long ago
(i.e. older than `PrivateClickMeasurement::maxAge`, which is currently 1 week).

* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::managerOrProxy):
(WebKit::NetworkSession::handlePrivateClickMeasurementConversion):
* Source/WebKit/NetworkProcess/PrivateClickMeasurement/PrivateClickMeasurementClient.h:
* Source/WebKit/NetworkProcess/PrivateClickMeasurement/PrivateClickMeasurementClientImpl.cpp:
(WebKit::PCM::ClientImpl::featureEnabled const):
(WebKit::PCM::ClientImpl::debugModeEnabled const):
(WebKit::PCM::ClientImpl::usesEphemeralDataStore const):

Enable the feature when using ephemeral sessions; however, add a new hook to return whether or not
we should actually attempt to persist entries to disk, or just keep them ephemerally in memory.

* Source/WebKit/NetworkProcess/PrivateClickMeasurement/PrivateClickMeasurementClientImpl.h:
* Source/WebKit/NetworkProcess/PrivateClickMeasurement/PrivateClickMeasurementDaemonClient.h:
* Source/WebKit/NetworkProcess/PrivateClickMeasurement/PrivateClickMeasurementEphemeralStore.cpp: Added.

Add a subclass of `PCM::Store` that represents an ephemeral, single-use attribution measurement that
doesn&apos;t rely on any on-disk presence.

(WebKit::PCM::EphemeralStore::insertPrivateClickMeasurement):
(WebKit::PCM::EphemeralStore::markAllUnattributedPrivateClickMeasurementAsExpiredForTesting):
(WebKit::PCM::EphemeralStore::attributePrivateClickMeasurement):
(WebKit::PCM::EphemeralStore::privateClickMeasurementToStringForTesting const):
(WebKit::PCM::EphemeralStore::allAttributedPrivateClickMeasurement):
(WebKit::PCM::EphemeralStore::markAttributedPrivateClickMeasurementsAsExpiredForTesting):
(WebKit::PCM::EphemeralStore::clearExpiredPrivateClickMeasurement):
(WebKit::PCM::EphemeralStore::clearPrivateClickMeasurement):
(WebKit::PCM::EphemeralStore::reset):
(WebKit::PCM::EphemeralStore::clearPrivateClickMeasurementForRegistrableDomain):
(WebKit::PCM::EphemeralStore::clearSentAttribution):
(WebKit::PCM::EphemeralStore::close):
* Source/WebKit/NetworkProcess/PrivateClickMeasurement/PrivateClickMeasurementEphemeralStore.h: Added.
(WebKit::PCM::EphemeralStore::create):
* Source/WebKit/NetworkProcess/PrivateClickMeasurement/PrivateClickMeasurementManager.cpp:
(WebKit::PrivateClickMeasurementManager::initializeStore const):
(WebKit::PrivateClickMeasurementManager::store):
(WebKit::PrivateClickMeasurementManager::store const):

Make this create either a `PersistentStore` or `EphemeralStore`, depending on whether or not we&apos;re
in an ephemeral session.

* Source/WebKit/NetworkProcess/PrivateClickMeasurement/PrivateClickMeasurementManager.h:
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/261507@main">https://commits.webkit.org/261507@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/89309216161f23d42edba664edf67616aff55ee9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/111912 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/21042 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/526 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/3674 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/120594 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/115974 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/22389 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/12115 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/86/builds/3420 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/117672 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/16617 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/99785 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/104930 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/98568 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/346 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/45602 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/13487 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/363 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/92803 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/14019 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/9773 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/19468 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52365 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8009 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/15968 "Built successfully") | | | 
| [❌ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/4363 "") | | | | 
<!--EWS-Status-Bubble-End-->

johnwilander commented on Source/WebKit/NetworkProcess/PrivateClickMeasurement/PrivateClickMeasurementManager.cpp:
```diff
@@ -336,6 +336,12 @@ void PrivateClickMeasurementManager::getSignedUnlinkableTokenForDestination(Sour
+        return completionHandler();
+    }
+
     store().insertPrivateClickMeasurement(WTFMove(measurement), type, WTFMove(completionHandler));
```
We should make sure store().insertPrivateClickMeasurement() release asserts if it's ever called in an ephemeral session.

whsieh replied:
(This comment actually made me realize that a much, much cleaner way to implement this is to refactor `PCM::Store` such that there's an ephemeral and persistent variant. I'm going to do that refactoring first, and that should help make this change a lot easier to reason about).

johnwilander commented on Source/WebKit/NetworkProcess/PrivateClickMeasurement/PrivateClickMeasurementManager.cpp:
```diff
@@ -440,37 +446,50 @@ void PrivateClickMeasurementManager::attribute(SourceSite&& sourceSite, Attribut
     if (!featureEnabled())
         return;
 
-    store().attributePrivateClickMeasurement(WTFMove(sourceSite), WTFMove(destinationSite), applicationBundleIdentifier, WTFMove(attributionTriggerData), m_isRunningTest ? WebCore::PrivateClickMeasurement::IsRunningLayoutTest::Yes : WebCore::PrivateClickMeasurement::IsRunningLayoutTest::No, [this, weakThis = WeakPtr { *this }] (auto attributionSecondsUntilSendData, auto debugInfo) {
```
Was this call dropped or moved?

whsieh replied:
This call was moved; however, I've since refactored the patch to not change core logic in `PrivateClickMeasurementManager`, but instead swap out the underlying data store for an ephemeral one.

johnwilander commented on Source/WebKit/NetworkProcess/PrivateClickMeasurement/PrivateClickMeasurementManager.h:
```diff
@@ -30,6 +30,7 @@
 #include "PrivateClickMeasurementClient.h"
 #include "PrivateClickMeasurementManagerInterface.h"
 #include "PrivateClickMeasurementStore.h"
+#include <WebCore/PrivateClickMeasurement.h>
```
Do we need this or can what's needed be forward declared?

whsieh replied:
Ah, good catch — this include was actually redundant altogether since it's already included by another one of the above headers.

johnwilander reviewed (commented):
Did you ask Alex how this will work with the daemon?

johnwilander commented on Source/WebKit/NetworkProcess/NetworkSession.cpp:
```diff
@@ -408,6 +408,12 @@ void NetworkSession::handlePrivateClickMeasurementConversion(WebCore::PCM::Attri
         auto firstPartyForCookies = redirectRequest.firstPartyForCookies();
 
+        bool hasAgedOut = WallTime::now() - ephemeralMeasurement.timeOfAdClick() > WebCore::PrivateClickMeasurement::maxAge();
+        if (hasAgedOut) {
```
Nice abstraction.

johnwilander commented on Source/WebKit/NetworkProcess/PrivateClickMeasurement/PrivateClickMeasurementClientImpl.cpp:
```diff
@@ -45,16 +45,17 @@ void ClientImpl::broadcastConsoleMessage(JSC::MessageLevel messageLevel, const S
-    return m_networkSession
-        && m_networkProcess->privateClickMeasurementEnabled()
-        && !m_networkSession->sessionID().isEphemeral();
+    return m_networkSession && m_networkProcess->privateClickMeasurementEnabled();
```
I remember Brady Eidson making use of how this was tied to ephemeral sessions for some unrelated things. Please check with him on what that was. It may have been regarding the special stateless NSURLSession PCM uses for sending attribution reports and how it relates to ephemeral sessions. It may also be the case that that issue is no longer relevant after the introduction of the daemon.

whsieh replied:
Interesting! I'll double check that — my understanding is that `PCM::NetworkLoader` already uses an ephemeral `NSURLSession` for reporting attribution (which should be compatible with our browsing session), but perhaps there's a subtlety here that I'm missing.

johnwilander commented on Source/WebKit/NetworkProcess/PrivateClickMeasurement/PrivateClickMeasurementEphemeralStore.cpp:
```diff
@@ -0,0 +1,145 @@
+void EphemeralStore::privateClickMeasurementToStringForTesting(CompletionHandler<void(String)>&& completionHandler) const
+{
+    if (!m_clickMeasurement)
+        return completionHandler("\nNo stored Private Click Measurement data.\n"_s);
```
Maybe add 'ephemeral' in this log statement to differentiate it from the persistent one.

johnwilander commented on Source/WebKit/NetworkProcess/PrivateClickMeasurement/PrivateClickMeasurementEphemeralStore.cpp:
```diff
@@ -0,0 +1,145 @@
+    completionHandler();
+}
+
+void EphemeralStore::clearPrivateClickMeasurement(CompletionHandler<void()>&& completionHandler)
```
Since ephemeral and persistent sessions can exist in parallel, please make sure that higher level calls to clear/reset PCM does so for both the ephemeral and the persistent store.

whsieh replied:
This comment made me realize that `clearPrivateClickMeasurement` _should_ probably be implemented even for the ephemeral data store, if only for robustness in the future. In practice, it doesn't really matter since we have the maxAge check *before* measuring and attributing the click anyways, but it's nicer to have parity with the persistent version.

johnwilander commented on Source/WebKit/NetworkProcess/PrivateClickMeasurement/PrivateClickMeasurementEphemeralStore.cpp:
```diff
@@ -0,0 +1,145 @@
+    completionHandler();
+}
+
+void EphemeralStore::reset()
```
Ditto.

johnwilander commented on Source/WebKit/NetworkProcess/PrivateClickMeasurement/PrivateClickMeasurementEphemeralStore.cpp:
```diff
@@ -0,0 +1,145 @@
+    m_clickMeasurement = std::nullopt;
+}
+
+void EphemeralStore::clearPrivateClickMeasurementForRegistrableDomain(WebCore::RegistrableDomain&& domain, CompletionHandler<void()>&& completionHandler)
```
Ditto.

johnwilander commented on Source/WebKit/NetworkProcess/PrivateClickMeasurement/PrivateClickMeasurementEphemeralStore.cpp:
```diff
@@ -0,0 +1,145 @@
+    m_clickMeasurement = WTFMove(attributionToClear);
+}
+
+void EphemeralStore::close(CompletionHandler<void()>&& completionHandler)
```
Ditto.

johnwilander commented on LayoutTests/http/tests/privateClickMeasurement/database-disabled-in-ephemeral-session.html:
```diff
@@ -8,7 +8,7 @@
 </head>
 <body onload="setTimeout(runTest, 0)">
 <div id="description">Tests that stored attributions cannot be converted in an ephemeral session.</div>
```
This needs to be changed too. "… doesn't use the data base in …" That'll effect text output, I think.

johnwilander commented on Source/WebKit/NetworkProcess/NetworkSession.cpp:
```diff
@@ -423,6 +429,9 @@ void NetworkSession::handlePrivateClickMeasurementConversion(WebCore::PCM::Attri
         return;
     }
 
+    if (m_sessionID.isEphemeral())
```
Is this just a safety measure? It feels strange to have below all of the handling of ephemeral measurement. Could it instead be an early check that if it's an ephemeral session and there is no pending ephemeral measurement, we return?

whsieh replied:
Hm...so I think we still want to early return here, otherwise we might end up with some confusing console messages about accepted triggering events, even though the ephemeral store would effectively ignore these attribution events. I'll move it up in this function, in the way you described.

johnwilander reviewed (approved):
OK to address comments in a follow-up patch but ideally in this one.

whsieh committed:
Refactor Private Click Measurement support for ephemeral sessions
https://bugs.webkit.org/show_bug.cgi?id=253600

Reviewed by John Wilander.

Refactor some support for the ephemeral version of Private Click Measurement (i.e., the version
currently used in Safari view controller), such that it's additionally compatible with ephemeral
browsing sessions.

* LayoutTests/http/tests/privateClickMeasurement/database-disabled-in-ephemeral-session-expected.txt: Renamed from LayoutTests/http/tests/privateClickMeasurement/conversion-disabled-in-ephemeral-session-expected.txt.
* LayoutTests/http/tests/privateClickMeasurement/database-disabled-in-ephemeral-session.html: Renamed from LayoutTests/http/tests/privateClickMeasurement/conversion-disabled-in-ephemeral-session.html.
* LayoutTests/http/tests/privateClickMeasurement/store-disabled-in-ephemeral-session-expected.txt: Removed.
* LayoutTests/http/tests/privateClickMeasurement/store-in-ephemeral-session-expected.txt: Added.
* LayoutTests/http/tests/privateClickMeasurement/store-in-ephemeral-session.html: Renamed from LayoutTests/http/tests/privateClickMeasurement/store-disabled-in-ephemeral-session.html.

Rename these tests to better reflect the new behavior in ephemeral sessions, and adjust the test
output to match.

* Source/WebCore/html/HTMLAnchorElement.cpp:
(WebCore::HTMLAnchorElement::parsePrivateClickMeasurement const):

Remove the early return here, and continue on call out to the client layer with PCM attribution
measurement data.

* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::willSendRedirectedRequestInternal):

Additionally avoid going through with attribution, if the click measurement happened too long ago
(i.e. older than `PrivateClickMeasurement::maxAge`, which is currently 1 week).

* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::managerOrProxy):
(WebKit::NetworkSession::handlePrivateClickMeasurementConversion):
* Source/WebKit/NetworkProcess/PrivateClickMeasurement/PrivateClickMeasurementClient.h:
* Source/WebKit/NetworkProcess/PrivateClickMeasurement/PrivateClickMeasurementClientImpl.cpp:
(WebKit::PCM::ClientImpl::featureEnabled const):
(WebKit::PCM::ClientImpl::debugModeEnabled const):
(WebKit::PCM::ClientImpl::usesEphemeralDataStore const):

Enable the feature when using ephemeral sessions; however, add a new hook to return whether or not
we should actually attempt to persist entries to disk, or just keep them ephemerally in memory.

* Source/WebKit/NetworkProcess/PrivateClickMeasurement/PrivateClickMeasurementClientImpl.h:
* Source/WebKit/NetworkProcess/PrivateClickMeasurement/PrivateClickMeasurementDaemonClient.h:
* Source/WebKit/NetworkProcess/PrivateClickMeasurement/PrivateClickMeasurementEphemeralStore.cpp: Added.

Add a subclass of `PCM::Store` that represents an ephemeral, single-use attribution measurement that
doesn't rely on any on-disk presence.

(WebKit::PCM::EphemeralStore::insertPrivateClickMeasurement):
(WebKit::PCM::EphemeralStore::markAllUnattributedPrivateClickMeasurementAsExpiredForTesting):
(WebKit::PCM::EphemeralStore::attributePrivateClickMeasurement):
(WebKit::PCM::EphemeralStore::privateClickMeasurementToStringForTesting const):
(WebKit::PCM::EphemeralStore::allAttributedPrivateClickMeasurement):
(WebKit::PCM::EphemeralStore::markAttributedPrivateClickMeasurementsAsExpiredForTesting):
(WebKit::PCM::EphemeralStore::clearExpiredPrivateClickMeasurement):
(WebKit::PCM::EphemeralStore::clearPrivateClickMeasurement):
(WebKit::PCM::EphemeralStore::reset):
(WebKit::PCM::EphemeralStore::clearPrivateClickMeasurementForRegistrableDomain):
(WebKit::PCM::EphemeralStore::clearSentAttribution):
(WebKit::PCM::EphemeralStore::close):
* Source/WebKit/NetworkProcess/PrivateClickMeasurement/PrivateClickMeasurementEphemeralStore.h: Added.
(WebKit::PCM::EphemeralStore::create):
* Source/WebKit/NetworkProcess/PrivateClickMeasurement/PrivateClickMeasurementManager.cpp:
(WebKit::PrivateClickMeasurementManager::initializeStore const):
(WebKit::PrivateClickMeasurementManager::store):
(WebKit::PrivateClickMeasurementManager::store const):

Make this create either a `PersistentStore` or `EphemeralStore`, depending on whether or not we're
in an ephemeral session.

* Source/WebKit/NetworkProcess/PrivateClickMeasurement/PrivateClickMeasurementManager.h:
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/261507@main
